pipeline/process_block.go

```diff
@@ -219,7 +257,115 @@ func (p *Pipeline) handleStepFinal(clock *pbsubstreams.Clock) error {
+		}
+
+	} else if blockData, _, err := execOutput.Get("sf.ethereum.type.v2.Block"); err == nil {
+		block := &pbeth.Block{}
+		if err := proto.Unmarshal(blockData, block); err != nil {
+			return fmt.Errorf("unmarshal block: %w", err)
+		}
+		block.TransactionTraces = block.TransactionTraces[p.partialProcessingState.processedTransactionsCount:]
+		p.partialProcessingState.processedTransactionsCount += uint64(len(block.TransactionTraces))
+		blockData, _ = proto.Marshal(block)
+		if err := execOutput.Set("sf.ethereum.type.v2.Block", blockData); err != nil {
+			return fmt.Errorf("set block: %w", err)
+		}
+	} else {
+		return fmt.Errorf("unsupported block type for partial blocks")
+	}
+
+	p.partialProcessingState.processedPartials = append(p.partialProcessingState.processedPartials, clock)
+
+	reqDetails := reqctx.Details(ctx)
+	if isBlockOverStopBlock(clock.Number, reqDetails.StopBlockNum) {
+		return nil // skip but don't take actions on a partial block for this
+	}
+	if p.isTier1 && p.checkPendingShutdown() {
+		return nil // skip but don't take actions on a partial block for this
+	}
+	if p.pendingUndoMessage != nil {
+		return nil
+	}
+	if !p.gate.shouldSendOutputs() {
+		return nil
+	}
+	if reqDetails.IsTier2Request {
+		panic("tier2 requests should never receive partial block")
+	}
+
+	if err := p.executeModules(ctx, execOutput, false, false); err != nil {
+		return fmt.Errorf("execute modules: %w", err)
+	}
+
+	mapModuleOutput := normalizeModuleOutput(p.mapModuleOutput, reqDetails.OutputModule)
+
+	if err = returnPartialDataOutput(clock, mapModuleOutput, p.respFunc, idx); err != nil {
+		return fmt.Errorf("failed to return module data output: %w", err)
+	}
+
+	return nil
+}
+
+func normalizeModuleOutput(in *pbsubstreamsrpc.MapModuleOutput, outputModule string) *pbsubstreamsrpc.MapModuleOutput {
+	if in == nil {
+		return &pbsubstreamsrpc.MapModuleOutput{
+			Name:      outputModule,
+			MapOutput: &anypb.Any{},
+		}
+	}
+	if in.Name == "" {
+		in.Name = outputModule
+	}
+	return in
+}
+
 func (p *Pipeline) handleStepNew(ctx context.Context, clock *pbsubstreams.Clock, cursor *bstream.Cursor, execOutput execout.ExecutionOutput, isFinalBlock bool) (err error) {
+
+	// if we get a 'new' block while handling partial blocks, we complete the missing partials based on its content.
+	// Ex: we got partials 3, 4, 7
+	//    -> partial 3 contains all transactions from partials 1+2+3
+	//    -> partial 4 only processes transactions from partial 4
+	//    -> when we get the full block (stepNEW), if it differs from the partial7's ID, we process it so that we send a 'partial 10' that contains transactions from partials 8,9,10
+	if reqctx.IncludePartialBlocks(ctx) { // this also covers 'partialBlocksOnly'
+		if p.partialProcessingState == nil || p.partialProcessingState.lastBlockID != clock.Id {
+			if err := p.handleStepPartial(ctx, clock, execOutput.Clone(), biggestPartialBlockIndex); err != nil { // we will reuse this execOutput so we clone it here
```

Comment:
Put the comment on the line before the `if`, I usually don't comment for this, but here, the line is really long so there is good chance the comment wouldn't even be visible in some editor.